### PR TITLE
refactor(compiler-cli): migrate source-map tests to new compliance tests (for inline templates)

### PR DESCRIPTION
**Pull request by petebacondarwin**

### CODE REVIEW

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+ * source file.
+ *
+ * The source-mappings are encoded into the expected source file in the form of an end-of-line
+ * comment that has the following syntax:
+ *
+ * ```
+ * <generated code> // SOURCE: "<path/to/original>" <original source>
+ * ```
+ *
+ * The `path/to/original` path will be relative to the source-map file.
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:23]: The commit message and readme mentions that `path/to/original` should be absolute, unlike here.

File: packages/compiler-cli/test/compliance/README.md [hunk 2]
```diff
@@ -126,6 +126,29 @@ are intelligently matched to check whether they are equivalent.
...
 
+### Source-map checks
+
+To check a mapping, add a `// SOURCE:` comment to the end of a line in an expectation file:
+
+```
+<generated code> // SOURCE: "<source-url>" <source code>
+```
+
+The generated code will be checked as usual by the `expectEmit()` helper.
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/README.md:137]: It looks like the check is done in `checkExpectations`, which strips the source mapping expectations before passing it into `expectEmit`, so I don't think this line is accurate (same for the commit message).

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+function unescape(str: string): string {
+  const replacements: Record<any, string> = {'\\n': '\n', '\\\\': '\\'};
+  return str.replace(/\\[n\\]/g, match => replacements[match]);
+}
+
+/**
+ * Process a generated file to extract human understandable segment mappings.
+ * These mappings are easier to compare in unit tests that the raw SourceMap mappings.
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:105]:
```suggestion
 * These mappings are easier to compare in unit tests than the raw SourceMap mappings.
```

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+ * @param fs the test file-system that holds the source and generated files.
+ * @param generatedFileName The name of the generated file to process.
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:107]:
```suggestion
 * @param generatedFileName The path of the generated file to process.
```

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+  const segments: SegmentMapping[] = [];
+  for (let i = 0; i < generatedFile.flattenedMappings.length; i++) {
+    const mapping = generatedFile.flattenedMappings[i];
+    const generatedStart = mapping.generatedSegment;
+    const next = generatedFile.flattenedMappings[i + 1];
+    const generatedEnd = next?.generatedSegment;
+    if (generatedEnd === undefined) {
+      continue;
+    }
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:130]: It looks like the loop should end one iteration earlier, in which case `next` and `generatedEnd` should no longer be possibly undefined.

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+    const originalFile = mapping.originalSource;
+    const originalStart = mapping.originalSegment;
+    let originalEnd = originalStart.next;
+    // Skip until we find an end segment that is after the start segment
+    while (originalEnd !== undefined && originalEnd.next !== originalEnd &&
+           originalEnd!.position === originalStart.position) {
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:137]: Non-null assertion shouldn't be necessary.

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    segments.push(segment);
+  }
+
+  return segments;
+}
+
+/**
+ * Check that the `expected` segment appears in the collection of `mappings`.
+ *
+ * An error is thrown with a useful message if the segment cannot be found.
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:158]: This is no longer the case

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+  }
+
+  return message.join('\n');
+}
+
+function prettyPrintMapping(mapping: SegmentMapping): string {
+  return [
+    '{',
+    `  generated: ${JSON.stringify(mapping.generated)}`,
+    `  source:    ${JSON.stringify(mapping.source)}`,
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:191]: The extra whitespace here may be misleading; perhaps aligning the colon to the right would help.

File: packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_for_templated_partial.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
+.ɵɵtext(0);\n  }\n\n  if (rf & 2) {\n    const item_r1 = ctx.$implicit;\n // SOURCE: "/ng_for_templated.ts" {{ item }}</ng-template>`
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_for_templated_partial.js:1]: The broader generated code is also quite different compared to full compile

File: packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_if_templated_partial.js [hunk 10]
```diff
@@ -0,0 +1,8 @@
+.ɵɵelementStart(0, "div");\n // SOURCE: "/ng_if_templated.ts" <div>
+…
+// TODO - map the bindings better
+// TODO: Work out how to fix the broken segment for the last item in a template
+.ɵɵelementEnd();\n // SOURCE: "/ng_if_templated.ts" </div>\n      
+…
+// TODO: the `ctx_r...` appears to be dependent upon previous tests!!!
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_if_templated_partial.js:7]: 😱

**Review by JoostK — COMMENTED** [review]
Type in last commit: "folllowed" in the second paragraph.

It's a bit sad that partial and full expectations have to split due the differences in source map information, but I don't have suggestions on improving that.

File: packages/compiler-cli/test/compliance/test_cases/test_case_schema.json [hunk 11]
```diff
@@ -20,10 +20,20 @@
...
-            "type": "boolean",
-            "default": false
+          "compilationModeFilter": {
+            "title": "An array of compilation modes under which this test-case should be run.",
+            "type": "array",
+            "items": {
+              "type": "string",
+              "enum": [
+                "full compile",
+                "partial compile + link"
```
[the remark this answers is not in the document]
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/test_cases/test_case_schema.json:30]: Happy to change it to `full compile` and `linked compile`.

File: packages/compiler-cli/test/compliance/README.md (hunk 2 above)
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/README.md:137]: Perhaps it is a bit poorly described.

What I mean is that the normal `expectEmit()` checks are still run on the code in the expected file, despite the `// SOURCE: ` comments stuff being processed and extracted.

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts (hunk 4 above)
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:107]: Well if we are going there then I probably need to change the parameter name too 😄

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts (hunk 8 above)
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:191]: OK

File: packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_for_templated_partial.js (hunk 9 above)
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_for_templated_partial.js:1]: What do you mean?

File: packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_if_templated_partial.js (hunk 10 above)
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_if_templated_partial.js:7]: Carried over from the original tests... I think that we have a module level counter in there somewhere.

File: packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_for_templated_partial.js (hunk 9 above)
Comment (JoostK) [reply packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_for_templated_partial.js:1]: So the full compile checks for

```js
i0.ɵɵtext(0) // SOURCE: "/ng_for_templated.ts" {{ item }}
```

whereas this is:

```js
.ɵɵtext(0);\n  }\n\n  if (rf & 2) {\n    const item_r1 = ctx.$implicit;\n // SOURCE: "/ng_for_templated.ts" {{ item }}</ng-template>`
```

File: packages/compiler-cli/test/compliance/README.md (hunk 2 above)
Comment (JoostK) [reply packages/compiler-cli/test/compliance/README.md:137]: Ah I see now. I interpreted this differently, but now that I read it again it makes sense.

File: packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts (hunk 4 above)
Comment (JoostK) [reply packages/compiler-cli/test/compliance/test_helpers/sourcemap_helpers.ts:107]: Works for me :-)

File: packages/compiler-cli/test/compliance/README.md (hunk 2 above)
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/README.md]: I tried to improve the wording.

File: packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_for_templated_partial.js (hunk 9 above)
Comment (petebacondarwin) [reply packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_for_templated_partial.js:1]: The difference is only in the source-mapping though, right?
The extra generated code that is shown in this expectation file does appear in the generated code for both compilation modes, it is just that it is not in the mapping for the full compile.

Comment (JoostK) [reply packages/compiler-cli/test/compliance/test_cases/source_mapping/inline_templates/ng_for_templated_partial.js:1]: Yes, it's only the source mapping. Nothing to worry about I guess, just an observation.

**Review by JoostK — APPROVED** [review]

**Pull request closed without merging**